### ISSUE TITLE
(wsd:on :close ws) message does not fire when browser closes.

### ISSUE THREAD

**Issue by knobo** [fukamachi/websocket-driver#23]
#21 fixes this.

**Comment by knobo** [comment]
after 4e65904bb2f2d7488cf8ea796188ef76c5a92940 I still don't see my log message when I reload my browser window. 
``` cl
(wsd:once :close ws (lambda (&rest args) (log:info "Closing connection"))
```

**Comment by knobo** [comment]
Even my pr does not invoke the :close message on close. So it does not fix it.

**Comment by knobo** [comment]
Looks like the :close message does not fire when client ends connection with 1002 protocol error.

**Comment by knobo** [comment]
You obviously want to call the callback in send

``` cl
(defmethod send :around ((ws ws) data &rest args)
  (when (eq (ready-state ws) :connecting)
    (return-from send
      (enqueue ws (cons data args))))

  (unless (eq (ready-state ws) :open)
    (when (getf args :callback)
       (funcall (getf args :callback)))
    (return-from send nil))

  (call-next-method))
```

**Comment by knobo** [comment]
Or, maybe just drop the `(send ..)` method call in  `send-close-frame`, and just call `close-connection` directly:

``` cl
(defun send-close-frame (ws reason code)
  (setf (ready-state ws) :closing)
  (close-connection ws reason code))
```

Or, since we are not sending a close-frame, then just do:

``` cl
(defmethod initialize-instance :after ((ws ws) &key)
  (setf (parser ws)
        (make-parser (ws-parse ws)
                     :require-masking (require-masking ws)
                     :max-length (max-length ws)
                     :message-callback
                     (lambda (message)
                       (emit :message ws message))
                     :ping-callback
                     (lambda (payload)
                       (send ws payload :type :pong))
                     :pong-callback
                     (lambda (payload)
                       (let ((callback (gethash payload (ping-callbacks ws))))
                         (when callback
                           (remhash payload (ping-callbacks ws))
                           (funcall callback))))
                     :close-callback
                     (lambda (data &key code)
                       (case (ready-state ws)
                         ;; closing request by the other peer
                         (:open
  (setf (ready-state ws) :closing)    ;;; This is new
  (close-connection ws reason code) ;;; And this
;; Instead of this:
;; (send-close-frame ws data code))
)
                         ;; probably the response for a 'close' frame
                         (otherwise
                          (close-connection ws data code)))
                       (setf (ws-stage (ws-parse ws)) 0))
                     :error-callback
                     (lambda (code reason)
                       (emit :error ws reason)
  (setf (ready-state ws) :closing)    ;;; This is new
  (close-connection ws reason code) ;;; And this
;; Instead of this:
;; (send-close-frame ws reason code)
                       (setf (ws-stage (ws-parse ws)) 0)))))
```

What do you think?

**Comment by knobo** [comment]
I've updated my pr #21

**Comment by knobo** [comment]
https://github.com/knobo/wstest  
For testing bugs relating to websocket (like this bug)

**Comment by knobo** [comment]
Maybe  this bug can be solved by solving #25
